python/lsst/rubintv/production/uploaders.py

```diff
@@ -0,0 +1,271 @@
+# This file is part of rubintv_production.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import json
+import os
+import time
+import logging
+
+from lsst.summit.utils.utils import dayObsIntToString
+
+from .channels import CHANNELS
+
+try:
+    from google.cloud import storage
+    from google.cloud.storage.retry import DEFAULT_RETRY
+    HAS_GOOGLE_STORAGE = True
+except ImportError:
+    HAS_GOOGLE_STORAGE = False
+
+__all__ = [
+    'Uploader',
+    'Heartbeater',
+]
+
+_LOG = logging.getLogger(__name__)
+
+
+class Uploader():
+    """Class for handling uploads to the Google Cloud Storage bucket.
+    """
+    HEARTBEAT_PREFIX = "heartbeats"
+
+    def __init__(self, bucketName):
+        if not HAS_GOOGLE_STORAGE:
+            from lsst.summit.utils.utils import GOOGLE_CLOUD_MISSING_MSG
+            raise RuntimeError(GOOGLE_CLOUD_MISSING_MSG)
+        self.client = storage.Client()
+        self.bucket = self.client.get_bucket(bucketName)
+        self.log = _LOG.getChild("googleUploader")
+
+    def uploadHeartbeat(self, channel, flatlinePeriod):
+        """Upload a heartbeat for the specified channel to the bucket.
+
+        Parameters
+        ----------
+        channel : `str`
+            The channel name.
+        flatlinePeriod : `float`
+            The period after which to consider the channel dead.
+
+        Returns
+        -------
+        success : `bool`
+            Did the upload succeed?
+        """
+        filename = "/".join([self.HEARTBEAT_PREFIX, channel]) + ".json"
```

Comment:
I'm guessing `os.path.join` is inappropriate because `self.HEARBEAT_PREFIX` is a non-file URI?

Reply:
Correct, these aren't paths, these are things to use the the GCS bucket.